Framework/CurveFitting/src/IMWDomainCreator.cpp

```diff
@@ -0,0 +1,494 @@
+    } else {
+      if (m_startX > m_endX) {
+        std::swap(m_startX, m_endX);
+      }
+      from = std::lower_bound(X.begin(), X.end(), m_startX);
+      to = std::upper_bound(from, X.end(), m_endX);
+    }
+  } else {
+    // x is descending
+    if (m_startX == EMPTY_DBL() && m_endX == EMPTY_DBL()) {
+      m_startX = X.front();
+      from = X.begin();
+      m_endX = X.back();
+      to = X.end();
+    } else if (m_startX == EMPTY_DBL() || m_endX == EMPTY_DBL()) {
+      throw std::invalid_argument(
+          "Both StartX and EndX must be given to set fitting interval.");
+    } else {
+      if (m_startX < m_endX) {
+        std::swap(m_startX, m_endX);
+      }
+      from = std::lower_bound(X.begin(), X.end(), m_startX, greaterIsLess);
+      to = std::upper_bound(from, X.end(), m_endX, greaterIsLess);
+    }
+  }
+
+  if (m_matrixWorkspace->isHistogramData()) {
+    if (X.end() == to) {
+      to = X.end() - 1;
+    }
+  }
+  return std::make_pair(std::distance(X.begin(), from),
+                        std::distance(X.begin(), to));
+}
+
+/**
+ * Return the size of the domain to be created.
+ */
+size_t IMWDomainCreator::getDomainSize() const {
+  setParameters();
+  size_t startIndex, endIndex;
+  std::tie(startIndex, endIndex) = getXInterval();
+  return endIndex - startIndex;
+}
+
+/**
+ * Initialize the function with the workspace.
+ * @param function :: Function to initialize.
+ */
+void IMWDomainCreator::initFunction(API::IFunction_sptr function) {
+  setParameters();
+  if (!function) {
+    throw std::runtime_error("Cannot initialize empty function.");
+  }
+  function->setWorkspace(m_matrixWorkspace);
+  function->setMatrixWorkspace(m_matrixWorkspace, m_workspaceIndex, m_startX,
+                               m_endX);
+  setInitialValues(*function);
+}
+
+/**
+ * Creates a workspace to hold the results. If the input workspace contains
+ * histogram data then so will this
+ * It assigns the X and input data values but no Y,E data for any functions
+ * @param nhistograms The number of histograms
+ * @param nyvalues The number of y values to hold
+ */
+API::MatrixWorkspace_sptr
+IMWDomainCreator::createEmptyResultWS(const size_t nhistograms,
+                                      const size_t nyvalues) {
+  size_t nxvalues(nyvalues);
+  if (m_matrixWorkspace->isHistogramData())
+    nxvalues += 1;
+  API::MatrixWorkspace_sptr ws = API::WorkspaceFactory::Instance().create(
+      "Workspace2D", nhistograms, nxvalues, nyvalues);
+  ws->setTitle("");
+  ws->setYUnitLabel(m_matrixWorkspace->YUnitLabel());
+  ws->setYUnit(m_matrixWorkspace->YUnit());
+  ws->getAxis(0)->unit() = m_matrixWorkspace->getAxis(0)->unit();
+  auto tAxis = new API::TextAxis(nhistograms);
+  ws->replaceAxis(1, tAxis);
+
+  const MantidVec &inputX = m_matrixWorkspace->readX(m_workspaceIndex);
+  const MantidVec &inputY = m_matrixWorkspace->readY(m_workspaceIndex);
+  const MantidVec &inputE = m_matrixWorkspace->readE(m_workspaceIndex);
```

Comment:
You can use the new `HistogramData` interface, e.g., `const auto &inputX = m_matrixWorkspace->x(index)`. Similar for Y and E. If you need a non-const reference (as below) use `mutableX(index)` etc.